File: src/Solutions/Euler159.hs
module Solutions.Euler159 where
import qualified Data.Set as Set
import Data.Array
import Utils.Array(funArray)
import Utils.Numeric(digitalRoot)

sumMdrs :: Int -> Int
sumMdrs bound =
    let mdrsArr :: Array Int Int
        mdrsArr = funArray 2 (bound - 1) mdrsFun

        mdrsFun :: Int -> Int
        mdrsFun n = foldr f (digitalRoot n) lsq
            where lsq = filter (\x -> n `mod` x == 0) $ takeWhile (\x -> x * x <= n) [2..]
                  f y x = max x (mdrsArr ! y + mdrsArr ! (n `div` y))
    in sum $ elems mdrsArr

euler159 :: IO String
euler159 = return $ show $ sumMdrs 1000000

File: rust/src/solutions/euler159.rs
use crate::utils::numeric::digits::Digits;

pub fn euler159() -> String {
    let mut mdrs_arr = Vec::new();
    for n in 2usize..1000000 {
        let mdrs_n = (2..)
            .take_while(|&x| x * x <= n)
            .filter(|&x| n % x == 0)
            .fold(n.digital_root(), |s, x| {
                std::cmp::max(s, mdrs_arr[x - 2] + mdrs_arr[n / x - 2])
            });
        mdrs_arr.push(mdrs_n);
    }
    mdrs_arr.into_iter().sum::<usize>().to_string()
}
